### PR TITLE
[Feature][Catalog] Add InMemoryCatalog for test and add new getCatalogTableFromConfig method

### PR DESCRIPTION
**Pull request by Hisoka-X**
## Purpose of this pull request

1. Add new `getCatalogTableFromConfig` method, so that we can directly get `CatalogTable` from source config. It will be used when refactor Catalog for move create `CatalogTable` logic to `TableSourceFactory` instance.
2. To prove new method can work fine, add test suite `InMemoryCatalog`, after this we can test catalog without any third party database.

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/seatunnel/tree/dev/docs/en/connector-v2)
  3. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  4. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Comment by Hisoka-X** [comment]
cc @ruanwenjun @hailin0

File: seatunnel-api/src/test/java/org/apache/seatunnel/api/table/catalog/CatalogTableUtilTest.java [hunk 1]
```diff
@@ -37,17 +36,13 @@
 import java.net.URISyntaxException;
 import java.net.URL;
 import java.nio.file.Paths;
+import java.util.List;
 
 public class CatalogTableUtilTest {
     @Test
     public void testSimpleSchemaParse() throws FileNotFoundException, URISyntaxException {
         String path = getTestConfigFile("/conf/simple.schema.conf");
-        Config config =
```
Comment (Hisoka-X) [inline seatunnel-api/src/test/java/org/apache/seatunnel/api/table/catalog/CatalogTableUtilTest.java:45]: ditto

File: seatunnel-api/src/test/java/org/apache/seatunnel/api/table/catalog/CatalogTableUtilTest.java [hunk 2]
```diff
@@ -61,12 +56,7 @@ public void testSimpleSchemaParse() throws FileNotFoundException, URISyntaxExcep
     @Test
     public void testComplexSchemaParse() throws FileNotFoundException, URISyntaxException {
         String path = getTestConfigFile("/conf/complex.schema.conf");
-        Config config =
```
Comment (Hisoka-X) [inline seatunnel-api/src/test/java/org/apache/seatunnel/api/table/catalog/CatalogTableUtilTest.java:64]: ditto

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/table/catalog/CatalogOptions.java [hunk 3]
```diff
@@ -56,11 +54,4 @@ public interface CatalogOptions {
                     .withDescription(
                             "The table names RegEx of the database to capture."
                                     + "The table name needs to include the database name, for example: database_.*\\.table_.*");
-
-    OptionRule.Builder BASE_RULE =
```
Comment (Hisoka-X) [inline seatunnel-api/src/main/java/org/apache/seatunnel/api/table/catalog/CatalogOptions.java:60]: I also removed some useless code.

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/table/catalog/CatalogTableUtil.java [hunk 4]
```diff
@@ -132,6 +135,54 @@ public static List<CatalogTable> getCatalogTables(Config config, ClassLoader cla
...
+     * schema, otherwise, return a catalog table with schema from catalog.
+     *
+     * @deprecated DO NOT invoke it in any new TableSourceFactory/TableSinkFactory, please directly
+     *     use TableSourceFactory/TableSinkFactory instance to get CatalogTable. We just use it to
+     *     transition the old CatalogTable creation logic. Details please <a
+     *     href="https://cwiki.apache.org/confluence/display/SEATUNNEL/STIP5-Refactor+Catalog+and+CatalogTable">check
+     *     </a>
+     */
+    @Deprecated
+    public static List<CatalogTable> getCatalogTablesFromConfig(
```
Comment (Hisoka-X) [inline seatunnel-api/src/main/java/org/apache/seatunnel/api/table/catalog/CatalogTableUtil.java:149]: after we delete `getCatalogTables` method, I will change this method name to `getCatalogTables`

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/table/catalog/CatalogTableUtil.java [hunk 5]
```diff
@@ -132,6 +135,54 @@ public static List<CatalogTable> getCatalogTables(Config config, ClassLoader cla
...
+                                List<CatalogTable> catalogTables =
+                                        catalog.getTables(readonlyConfig);
+                                log.info(
+                                        String.format(
+                                                "Get catalog tables, cost time: %d",
+                                                System.currentTimeMillis() - startTime));
+                                return catalogTables;
+                            }
+                        })
+                .orElse(Collections.emptyList());
```
Comment (ruanwenjun) [inline seatunnel-api/src/main/java/org/apache/seatunnel/api/table/catalog/CatalogTableUtil.java]: Do we need to throw exception if the Catalog is not exist or catalogTable is empty.

File: seatunnel-api/src/test/java/org/apache/seatunnel/api/table/catalog/CatalogTableUtilTest.java (hunk 1 above)
Comment (ruanwenjun) [reply seatunnel-api/src/test/java/org/apache/seatunnel/api/table/catalog/CatalogTableUtilTest.java:45]: After remove this ,can we replace the system properties in config file?

Comment (Hisoka-X) [reply seatunnel-api/src/test/java/org/apache/seatunnel/api/table/catalog/CatalogTableUtilTest.java:45]: No, but it useless on this test case. Without the removed code, the test also will pass.

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/table/catalog/CatalogTableUtil.java (hunk 5 above)
Comment (Hisoka-X) [reply seatunnel-api/src/main/java/org/apache/seatunnel/api/table/catalog/CatalogTableUtil.java]: Sounds reasonable, let me add it.

Comment (Hisoka-X) [reply seatunnel-api/src/main/java/org/apache/seatunnel/api/table/catalog/CatalogTableUtil.java]: Addressed.

File: seatunnel-api/src/test/java/org/apache/seatunnel/api/table/catalog/CatalogTableUtilTest.java (hunk 1 above)
Comment (ruanwenjun) [reply seatunnel-api/src/test/java/org/apache/seatunnel/api/table/catalog/CatalogTableUtilTest.java:45]: OK, this is only used in test.

**Review by ruanwenjun — DISMISSED** [review]
LGTM

**Review by Carl-Zhou-CN — APPROVED** [review]
+1

File: seatunnel-api/src/test/java/org/apache/seatunnel/api/table/catalog/InMemoryCatalog.java [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+import com.google.common.collect.Lists;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class InMemoryCatalog implements Catalog {
+    private static final Logger LOGGER = LoggerFactory.getLogger(InMemoryCatalog.class);
```
Comment (liugddx) [inline seatunnel-api/src/test/java/org/apache/seatunnel/api/table/catalog/InMemoryCatalog.java]: use @Slf4j?

Comment (Hisoka-X) [reply seatunnel-api/src/test/java/org/apache/seatunnel/api/table/catalog/InMemoryCatalog.java]: done

**Comment by liugddx** [comment]
Is this not triggering CI?

**Comment by Hisoka-X** [comment]
> Is this not triggering CI?

Triggered. But not async with main repo. Let me investigate it.
image

**Review by liugddx — APPROVED** [review]
+1

**Review by ruanwenjun — APPROVED** [review]
+1

**Review by hailin0 — APPROVED** [review]

**Comment by Hisoka-X** [comment]
Thanks @liugddx @hailin0 @Carl-Zhou-CN @ruanwenjun

**Pull request merged**